crypto/chacha/asm/chacha-armv8-sve.pl

```diff
@@ -0,0 +1,1008 @@
+    $code .= "\t$opcode\t".join(',',@_,$arg)."\n";
+}
+
+my ($outp,$inp,$len,$key,$ctr) = map("x$_",(0..4));
+my ($ictr) = ("w4");
+my ($state) = ("x5");
+my ($veclen,$blocks) = ("x6","x7");
+my ($saved_outp) = ("x8");
+my @mx=map("z$_",(0..7,16..23));
+my ($xa0,$xa1,$xa2,$xa3, $xb0,$xb1,$xb2,$xb3,
+    $xc0,$xc1,$xc2,$xc3, $xd0,$xd1,$xd2,$xd3) = @mx;
+my @xt=map("z$_",(24..31,8..11));
+my ($rot8) = ("z12");
+my ($xt0,$xt1,$xt2,$xt3,$xt4,$xt5,$xt6,$xt7,$xt8,$xt9,$xt10,$xt11)=@xt;
+my $debug_encoder=0;
+
+sub SVE_ADD() {
+	my $unpred_mode = shift;
+	my $x = shift;
+	my $y = shift;
+
+	if ($unpred_mode == 1) {
+$code.=<<___;
+	add	@mx[$x].s,@mx[$x].s,@mx[$y].s
+___
+	} else {
+$code.=<<___;
+	add	@mx[$x].s,p0/m,@mx[$x].s,@mx[$y].s
+___
+	}
+	if (@_) {
+		&SVE_ADD($unpred_mode,@_);
+	}
+}
+
+sub SVE_EOR() {
+	my $unpred_mode = shift;
+	my $x = shift;
+	my $y = shift;
+
+	if ($unpred_mode == 1) {
+$code.=<<___;
+	eor	@mx[$x].d,@mx[$x].d,@mx[$y].d
+___
+	} else {
+$code.=<<___;
+	eor	@mx[$x].s,p0/m,@mx[$x].s,@mx[$y].s
+___
+	}
+	if (@_) {
+		&SVE_EOR($unpred_mode,@_);
+	}
+}
+
+sub SVE_LSL() {
+	my $unpred_mode = shift;
+	my $bits = shift;
+	my $x = shift;
+	my $y = shift;
+	my $next = $x + 1;
+
+$code.=<<___;
+	lsl	@xt[$x].s,@mx[$y].s,$bits
+___
+	if (@_) {
+		&SVE_LSL($unpred_mode,$bits,$next,@_);
+	}
+}
+
+sub SVE_LSR() {
+	my $unpred_mode = shift;
+	my $bits = shift;
+	my $x = shift;
+
+	if ($unpred_mode == 1) {
+$code.=<<___;
+	lsr	@mx[$x].s,@mx[$x].s,$bits
+___
+	} else {
+$code.=<<___;
+	lsr	@mx[$x].s,p0/m,@mx[$x].s,$bits
+___
+	}
+	if (@_) {
+		&SVE_LSR($unpred_mode,$bits,@_);
+	}
+}
+
+sub SVE_ORR() {
+	my $unpred_mode = shift;
+	my $x = shift;
+	my $y = shift;
+	my $next = $x + 1;
+
+	if ($unpred_mode == 1) {
+$code.=<<___;
+	orr	@mx[$y].d,@mx[$y].d,@xt[$x].d
+___
+	} else {
+$code.=<<___;
+	orr	@mx[$y].s,p0/m,@mx[$y].s,@xt[$x].s
+___
+	}
+	if (@_) {
+		&SVE_ORR($unpred_mode,$next,@_);
+	}
+}
+
+sub SVE_REV16() {
+	my $x = shift;
+
+$code.=<<___;
+	revh	@mx[$x].s,p0/m,@mx[$x].s
+___
+	if (@_) {
+		&SVE_REV16(@_);
+	}
+}
+
+sub SVE_ROT8() {
+	my $x = shift;
+
+$code.=<<___;
+	tbl	@mx[$x].b,{@mx[$x].b},$rot8.b
+___
+	if (@_) {
+		&SVE_ROT8(@_);
+	}
+}
+
+sub SVE2_XAR() {
+	my $bits = shift;
+	my $x = shift;
+	my $y = shift;
+	my $rbits = 32-$bits;
+
+$code.=<<___;
+	xar	@mx[$x].s,@mx[$x].s,@mx[$y].s,$rbits
+___
+	if (@_) {
+		&SVE2_XAR($bits,@_);
+	}
+}
+
+sub SVE_QR_GROUP() {
+	my $have_sve2 = shift;
+	my $unpred_mode = shift;
+	my ($a0,$b0,$c0,$d0,$a1,$b1,$c1,$d1,$a2,$b2,$c2,$d2,$a3,$b3,$c3,$d3) = @_;
+
+	&SVE_ADD($unpred_mode,$a0,$b0,$a1,$b1,$a2,$b2,$a3,$b3);
+	&SVE_EOR($unpred_mode,$d0,$a0,$d1,$a1,$d2,$a2,$d3,$a3);
+	&SVE_REV16($d0,$d1,$d2,$d3);
+
+	&SVE_ADD($unpred_mode,$c0,$d0,$c1,$d1,$c2,$d2,$c3,$d3);
+	if ($have_sve2 == 0) {
+		&SVE_EOR($unpred_mode,$b0,$c0,$b1,$c1,$b2,$c2,$b3,$c3);
+		&SVE_LSL($unpred_mode,12,0,$b0,$b1,$b2,$b3);
+		&SVE_LSR($unpred_mode,20,$b0,$b1,$b2,$b3);
+		&SVE_ORR($unpred_mode,0,$b0,$b1,$b2,$b3,);
```

Comment:
Any reason not to use SRI here?
https://developer.arm.com/documentation/ddi0602/2022-03/SVE-Instructions/SRI--Shift-right-and-insert--immediate--?lang=en